src/interface/meshblock_data.cpp

```diff
@@ -72,124 +82,129 @@ void MeshBlockData<T>::Add(const std::string &label, const Metadata &metadata) {
-      } else {
-        newvar = sv->second;
+
+      auto fv = face_map.find(name);
+      if (fv != face_map.end()) {
+        PARTHENON_REQUIRE_THROWS(!found, "MeshBlockData::CopyFrom: Variable '" + name +
+                                             "' found more than once");
+        found = true;
+        add_var(fv->second);
       }
-      Add(newvar);
-    }
-    auto fv = face_map.find(name);
-    if (fv != face_map.end()) {
-      if (found) {
-        std::stringstream msg;
-        msg << "MeshBlockData: " << name << " found more than once!" << std::endl;
-        PARTHENON_THROW(msg);
+
+      if (!found && (resolved_packages_ != nullptr)) {
+        // check if this is a sparse base name, if so we get its pool of sparse_ids,
+        // otherwise we get an empty pool
+        const auto &sparse_pool = resolved_packages_->GetSparsePool(name);
+
+        // add all sparse ids of the pool
+        for (const auto iter : sparse_pool.pool()) {
+          // this variable must exist, if it doesn't something is very wrong
+          const auto &v = varMap_.at(MakeVarLabel(name, iter.first));
+          add_var(v);
+          found = true;
+        }
       }
-      found = true;
-      Add(fv->second);
-    }
-    if (!found) {
-      std::stringstream msg;
-      msg << "MeshBlockData: " << name << " found more than once!" << std::endl;
-      PARTHENON_THROW(msg);
+
+      PARTHENON_REQUIRE_THROWS(found, "MeshBlockData::CopyFrom: Variable '" + name +
+                                          "' not found");
     }
   }
 }
+
+// Constructor for getting sub-containers
+// the variables returned are all shallow copies of the src container.
+// Optionally extract only some of the sparse ids of src variable.
+template <typename T>
+MeshBlockData<T>::MeshBlockData(const MeshBlockData<T> &src,
+                                const std::vector<std::string> &names) {
+  CopyFrom(src, true, names);
+}
+
 template <typename T>
 MeshBlockData<T>::MeshBlockData(const MeshBlockData<T> &src,
                                 const std::vector<MetadataFlag> &flags) {
-  auto var_map = src.GetCellVariableMap();
-  auto sparse_map = src.GetSparseMap();
-  auto face_map = src.GetFaceMap();
-  for (auto &it : var_map) {
-    auto n = it.first;
-    auto v = it.second;
-    if (v->metadata().AnyFlagsSet(flags)) {
-      Add(v);
-    }
-  }
-  for (auto &it : sparse_map) {
-    auto n = it.first;
-    auto v = it.second;
-    if (v->metadata().AnyFlagsSet(flags)) {
-      Add(v);
-    }
-  }
-  for (auto &it : face_map) {
-    auto n = it.first;
-    auto v = it.second;
-    if (v->metadata().AnyFlagsSet(flags)) {
-      Add(v);
-    }
-  }
+  CopyFrom(src, true, {}, flags);
 }
 
 // provides a container that has a single sparse slice
 template <typename T>
-std::shared_ptr<MeshBlockData<T>> MeshBlockData<T>::SparseSlice(int id) {
+std::shared_ptr<MeshBlockData<T>> MeshBlockData<T>::SparseSlice(int sparse_id) {
```

Comment:
Is this function used downstream on your side a lot?
My brain says it's kind of counterintuitive that a "sparse slice" contains *all* dense fields.

Reply:
Good question. I don't know if it's being used but the old behavior was to include all dense fields, so I kept that. From a practical point of view, especially considering that sparse ids would often be used for different materials, I think it makes sense to include all dense fields. Quantities that are tied to specific materials would be sparse with a specific sparse id, quantities that apply to all materials would be dense. So taking a sparse slice is essentially selecting a subset of materials, but all the quantities that are not tied to materials (i.e. dense fields) should still be included, because they will probably be needed.

That being said, I'm happy to rename or remove this function if it's not actually being used. Also, there is redundancy between this function and the more general `CopyFrom` anyway. In fact, I'll have this function call `CopyFrom` to reduce code duplication.